vizro-core/tests/unit/vizro/models/test_page.py

```diff
@@ -101,6 +101,8 @@ def test_action_auto_generation_valid(self, standard_px_chart):
 # TODO: Add unit tests for private methods in page build
 def test_page_build_left_side_removed(standard_px_chart):
     page = vm.Page(title="Single Page", components=[vm.Graph(id="scatter_chart", figure=standard_px_chart)])
-    vm.Dashboard(pages=[page]).build()
+    dashboard = vm.Dashboard(pages=[page])
+    dashboard.pre_build()
+    dashboard.navigation.pre_build()
```

Comment:
@antonymilne - I've implemented the changes we've discussed and I think they are an improvement to what we've had before (Navigation being dependent on `Dashboard.build()`). 

However, this set-up still bugs me 😄 It's related to the fact that we build the navigation in the Page, and the navigation now depends on the dashboard.pre_build(). So this page.build() test currently requires a pre-build of the dashboard and the navigation (otherwise it cannot find the private _selector attribute).

I think for now that's fine, as these steps will always be executed before the `page.build()`. However, I am sure we can solve this by moving the `navigation.pre-build()/build()` to the dashboard. Just wanted to emphasize above as it plays into our discussion enabling arguments inside the `build()` methods.

Reply:
Agree with all this. There's actually one potentially problem which we should check: does `dashboard.pre_build` need to be run before `navigation.pre_build` in order to work? Or if you swap those round is it still ok?

Would things be neater if we moved the registration of pages into a `Dashboard` validator so that it's executed before any pre-build? We should think carefully before doing that though because it would be very easy to have a contaminated Dash page registry unless we empty it out first (same problem as we have with the model manager).

Reply:
The order here doesn't matter, the test will still pass if we switch the order. I think it's because we are re-using the same validator in the Navigation and Accordion, so either way the pages will get auto-populated if no Navigation is provided.

Given the order above doesn't matter - shall we leave as is or shall I try to move it to a validator? We actually had a problem of contaminated dash page registry leading to test interdependence which was fixed in this PR: https://github.com/mckinsey/vizro/pull/84

However, I had to remove each page individually as initiating with an empty OrderedDict again left the dash.page_registry permanently empty.

Reply:
**EDIT:** Seems like we can't move the registration of pages into a Dashboard validator, you'll get the error `"dash.exceptions.PageError: `dash.register_page()` must be called after app instantiation"`. So I guess it can only live in the pre-build or build?

![Screenshot 2023-10-04 at 18 19 54](https://github.com/mckinsey/vizro/assets/90609403/fdf9dc04-556a-4ba4-b617-beffe552e317)

Reply:
Ah yes, that is a problem. Possibly we should actually treat the `dash.register_page` as something quite special that lives in `Vizro.build` itself:
```
    def build(self, dashboard: Dashboard):
        for page in dashboard.pages:
             dash.register_page(...)
        self._pre_build()
        ....
```

Let's not do it now, given that it seems to work ok in `pre_build` and the order doesn't seem to matter.

This does feel a bit fragile though, and I can imagine it inadvertently introducing bugs in future - we always need to stay careful to make sure that `pre_build` methods can run in any order with the same result, and that is easy to forget. There seem to be a few problems with page registration and model manager anyway (e.g. #59), so let's bear in mind the possibility of moving this bit to `Vizro.build` when thinking about those since it might improve things overall.

I have some comments on #84 but let's save that for another conversation 🙂

Reply:
I have a few questions on that one, but I'll add it to the list of minor topics to catch-up on! Will leave this conversation open, so we can go back :)